File: test/_harnesses/typescript/broken.ts
import { getDataView, writeBuffer, runTest } from "./util";

import * as BrokenMessages from '../../../out/generated/typescript/BrokenMessages';

const broken = new BrokenMessages.TruncatedMessage();
broken.x = 1.0;
broken.y = 2.0;

function generate(filePath: string, softAssert: (condition: boolean, label: string) => void) {
    const data = new ArrayBuffer(broken.getSizeInBytes());
    const da = new BrokenMessages.DataAccess(data);
    broken.writeBytes(da, false);

    writeBuffer(Buffer.from(data, 0, da.currentOffset), filePath);

    softAssert(broken.getSizeInBytes() == da.currentOffset, "written bytes check");
}

function read(filePath: string, softAssert: (condition: boolean, label: string) => void) {
    const dv = getDataView(filePath);
    let errMsg: string;
    try {
        const input = BrokenMessages.FullMessage.fromBytes(dv);
    }
    catch (e) {
        errMsg = e.message;
    }

    softAssert(errMsg === "Could not read FullMessage from offset 8 (RangeError -- Offset is outside the bounds of the DataView)", "reading broken message");
}


runTest(generate, read);

File: test/_harnesses/csharp/broken.cs
using System;
using System.IO;
using System.Collections.Generic;

using BrokenMessages;

class BrokenHarness: TestHarness {

    static void Main(string[] args) {
        Dictionary<string, string> parsedArgs = parseArguments(args);

        var broken = new BrokenMessages.TruncatedMessage();
        broken.x = 1.0f;
        broken.y = 2.0f;

        if (parsedArgs.ContainsKey("generate"))
        {
            string outPath = parsedArgs["generate"];
            string outDir = System.IO.Path.GetDirectoryName(outPath);
            System.IO.Directory.CreateDirectory(outDir);
            FileStream f = new FileStream(outPath, FileMode.Create);
            BinaryWriter bw = new BinaryWriter(f);
            broken.WriteBytes(bw, false);

            softAssert(broken.GetSizeInBytes() == bw.BaseStream.Position, "written bytes check");
        }
        else if (parsedArgs.ContainsKey("read"))
        {
            FileStream f = File.OpenRead(parsedArgs["read"]);
            BinaryReader br = new BinaryReader(f);
            string errMsg = "";
            try
            {
                var input = BrokenMessages.FullMessage.FromBytes(br);
            }
            catch (BrokenMessages.DataReadErrorException e)
            {
                errMsg = e.Message;
            }
            softAssert(errMsg == "Could not read FullMessage from offset 8", "reading broken message");
        }


        check();
    }
}
